File: iosApp/iosApp/Sources/ViewModels/FavoritesViewModel.swift
import SwiftUI
import Shared

@MainActor
class FavoritesViewModel: ObservableObject {
    @Published var favorites: [Character_] = []
    @Published var isLoading = false
    @Published var error: String?

    private let getFavoritesUseCase = KoinHelper.shared.resolve(GetFavoritesUseCase.self)
    private let toggleFavoriteUseCase = KoinHelper.shared.resolve(ToggleFavoriteUseCase.self)

    func observeFavorites() async {
        isLoading = true
        do {
            for try await list in getFavoritesUseCase.invoke() {
                let chars = list as? [Character_] ?? []
                favorites = chars
                isLoading = false
            }
        } catch {
            self.error = error.localizedDescription
            isLoading = false
        }
    }

    func removeFavorite(character: Character_) async {
        do {
            let _ = try await toggleFavoriteUseCase.invoke(characterId: character.id)
        } catch {
            self.error = error.localizedDescription
        }
    }
}

File: composeApp/src/androidMain/kotlin/com/example/kmp_basic_app/viewmodel/FavoritesViewModel.kt
package com.example.kmp_basic_app.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.kmp_basic_app.domain.model.Character
import com.example.kmp_basic_app.domain.usecase.GetFavoritesUseCase
import com.example.kmp_basic_app.domain.usecase.ToggleFavoriteUseCase
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class FavoritesUiState(
    val favorites: List<Character> = emptyList(),
    val isLoading: Boolean = true
)

class FavoritesViewModel(
    private val getFavoritesUseCase: GetFavoritesUseCase,
    private val toggleFavoriteUseCase: ToggleFavoriteUseCase
) : ViewModel() {

    private val _uiState = MutableStateFlow(FavoritesUiState())
    val uiState: StateFlow<FavoritesUiState> = _uiState.asStateFlow()

    init {
        observeFavorites()
    }

    private fun observeFavorites() {
        viewModelScope.launch {
            getFavoritesUseCase().collect { favorites ->
                _uiState.update {
                    it.copy(
                        favorites = favorites.map { char -> char.copy(isFavorite = true) },
                        isLoading = false
                    )
                }
            }
        }
    }

    fun removeFavorite(character: Character) {
        viewModelScope.launch {
            toggleFavoriteUseCase(character)
        }
    }
}
